### PR TITLE
Add documentation on Marshal and Unmarshal

### PR DESCRIPTION
**Pull request by R-maan**

### CODE REVIEW

File: ion/unmarshal.go [hunk 1]
```diff
@@ -31,6 +31,58 @@ var (
...
+//		ion.Decimal/interface{}							decimal
+//		ion.Timestamp/interface{}						timestamp
+//		string/interface{}								symbol
+//		string/interface{}								string
+//		string/interface{}								string
+//		[]byte/[]interface{}{}							clob
+//		[]byte/[]interface{}{}							blob
+//		[]interface{}{}									list
+//		[]interface{}{}									sexp
+//		map[string]interface{}{}/struct/interface{}		struct
```
Comment (R-maan) [inline ion/unmarshal.go:86]: I'll fix the spacing

File: ion/marshal.go [hunk 2]
```diff
@@ -38,6 +38,42 @@ type Marshaler interface {
 }
 
 // MarshalText marshals values to text ion.
+//
+// Different Go types can be passed to be Marshalled to their corresponding Ion types.
```
Comment (byronlin-bq) [inline ion/marshal.go:42]: marshaled

Comment (byronlin13) [inline ion/marshal.go]:
```suggestion
// Different Go types can be passed into MarshalText() to be marshalled to their corresponding Ion types. e.g.,
```

File: ion/marshal.go [hunk 3]
```diff
@@ -38,6 +38,42 @@ type Marshaler interface {
...
+//
+//	val, err := MarshalText(9)
```
Comment (byronlin13) [inline ion/marshal.go]: Theo mentioned godoc will convert doc to codeblock if it is prepended with 4 spaces

https://github.com/amzn/ion-hash-go/pull/50#discussion_r462739798

File: ion/marshal.go [hunk 4]
```diff
@@ -38,6 +38,42 @@ type Marshaler interface {
...
+//		t.Fatal(err)
+//	}
+//
+// 	type root struct {
+//		A struct {
+//			B int `ion:"b"`
+//		}
+//		C int `ion:"c"`
+//	}
+//  v := root{A: {B: 6}, C: 7}
```
Comment (byronlin13) [inline ion/marshal.go]: This is not indented properly

File: ion/marshal.go [hunk 5]
```diff
@@ -38,6 +38,42 @@ type Marshaler interface {
...
+// Should the value for marshalling needs to have annotations, it must be wrapped in a
+// Go struct with exactly 2 fields, where the other field of the struct is a slice of
+// string and tagged `ion:",annotations"`, and this field can carry all the desired
+// annotations.
+//
+//	type foo struct {
+//		Value   int
+//		AnyName []string `ion:",annotations"`
+//	}
+//  v := foo{5, []string{"some", "annotations"}}   //some::annotations::5
```
Comment (byronlin13) [inline ion/marshal.go:75]: Not indented properly

File: ion/marshal.go [hunk 6]
```diff
@@ -38,6 +38,42 @@ type Marshaler interface {
...
+//	}
+//
+// 	type root struct {
+//		A struct {
+//			B int `ion:"b"`
+//		}
+//		C int `ion:"c"`
+//	}
+//  v := root{A: {B: 6}, C: 7}
+//	val, err := MarshalText(v)
```
Comment (byronlin13) [inline ion/marshal.go]: Can we do fmt.Println(val) to show the Ion text

File: ion/marshal.go [hunk 7]
```diff
@@ -38,6 +38,42 @@ type Marshaler interface {
...
+//	if err != nil {
+//		t.Fatal(err)
+//	}
+//
+//
+// Should the value for marshalling needs to have annotations, it must be wrapped in a
```
Comment (byronlin13) [inline ion/marshal.go]:
```suggestion
// Should the value for marshalling require annotations, it must be wrapped in a
```

File: ion/unmarshal.go [hunk 8]
```diff
@@ -31,6 +31,58 @@ var (
 )
 
 // Unmarshal unmarshals Ion data to the given object.
+//
+// Passing proper type for the object in which the data is  supposed to
```
Comment (byronlin13) [inline ion/unmarshal.go]:
```suggestion
// User must pass the proper object type to the unmarshalled Ion data.
```

File: ion/unmarshal.go [hunk 9]
```diff
@@ -31,6 +31,58 @@ var (
...
+// unmarshalled to is important. Below is the mapping between Golang
```
Comment (byronlin13) [inline ion/unmarshal.go]: Lets be consistent with `Go` and `Golang`

File: ion/unmarshal.go [hunk 10]
```diff
@@ -31,6 +31,58 @@ var (
...
+// native type and Ion types.
```
Comment (byronlin13) [inline ion/unmarshal.go]:
```suggestion
// native type and Ion types. e.g.,
```

File: ion/unmarshal.go [hunk 11]
```diff
@@ -31,6 +31,58 @@ var (
...
+//	  t.Fatal(err)
+// }
+//
+// err = UnmarshalString(true, &boolVal)
+// if err != nil {
+//	  t.Fatal(err)
+// }
+//
+//
+// To unmarshal an Ion value with annotations, the object sent to Unmarshal
```
Comment (byronlin13) [inline ion/unmarshal.go]:
```suggestion
// To unmarshal an Ion value with annotations, the object passed to Unmarshal
```

File: ion/unmarshal.go [hunk 12]
```diff
@@ -31,6 +31,58 @@ var (
...
+//  	Go native type   								Ion Type
+//	--------------------------						---------------
+//		nil/interface{}									null
+//		bool/interface{}								bool
+//		Any ints/uints/big.Int/interface{}				int
+//		float32/float64/interface{}						float
+//		ion.Decimal/interface{}							decimal
+//		ion.Timestamp/interface{}						timestamp
+//		string/interface{}								symbol
+//		string/interface{}								string
```
Comment (byronlin13) [inline ion/unmarshal.go:81]: duplicate

Comment (R-maan) [reply ion/unmarshal.go:81]: removed

File: ion/marshal.go (hunk 6 above)
Comment (R-maan) [reply ion/marshal.go]: great idea, added

File: ion/unmarshal.go [hunk 13]
```diff
@@ -31,6 +31,60 @@ var (
...
+//
+//     var val foo
+//     err := UnmarshalString("age::10", &val)
+//     if err != nil {
+//         t.Fatal(err)
+//     }
+//     fmt.Println(val) // prints out: {10 [age]}
+//
+//  	Go native type   								Ion Type
+//	--------------------------						---------------
```
Comment (byronlin13) [inline ion/unmarshal.go:73]: hmm it seems the ion types aren't aligned

**Comment by R-maan** [comment]
Closing as it is merged into main repo

**Pull request closed without merging**